NV_low_latency2 Impossible to use `timings` array after mutably borrowing in `GetLatencyMarkerInfoNV`

CC @repi @VZout, in case you run into this while incorporating the updated extension bindings in your code. This happens after https://github.com/ash-rs/ash/pull/816 was merged.

The `timings()` builder function on `GetLatencyMarkerInfoNV` mutably borrows a slice of `LatencyTimingsFrameReportNV`, which itself contains pointers and a lifetime. Rust seems to assume

```rust
#[test]
fn lifetime() {
    let mut timings = vec![vk::LatencyTimingsFrameReportNV::default()];
    let info = vk::GetLatencyMarkerInfoNV::default().timings(&mut timings);
    drop(info); // Lifetime wouldn't extend beyond this point if info is no longer used
    dbg!(timings);
}
```

Results in:

```
error[E0505]: cannot move out of `timings` because it is borrowed
  --> ash/src/extensions/nv/low_latency2.rs:93:5
   |
90 |     let mut timings = vec![vk::LatencyTimingsFrameReportNV::default()];
   |         ----------- binding `timings` declared here
91 |     let info = vk::GetLatencyMarkerInfoNV::default().timings(&mut timings);
   |                                                              ------------ borrow of `timings` occurs here
92 |     drop(info); // Lifetime wouldn't extend beyond this point if info is no longer used
93 |     dbg!(timings);
   |     ^^^^^^^^^^^^^
   |     |
   |     move out of `timings` occurs here
   |     borrow later used here
   |
   = note: this error originates in the macro `dbg` (in Nightly builds, run with -Z macro-backtrace for more info)
```

(EDIT: Instead of `drop(info);`, we could also wrap `let info` in a new lifetime scope `{}`)

More context: https://github.com/ash-rs/ash/pull/815#discussion_r1392679017

_Originally posted by @MarijnS95 in https://github.com/ash-rs/ash/pull/816#discussion_r1405804943_

With this definition:
```rs
pub fn timings(mut self, timings: &'a mut [LatencyTimingsFrameReportNV<'_>]) -> Self {
```

Correct use compiles fine, and I get sensible errors for both incorrect use patterns:

`timings` ref-after-free:
```
error[E0597]: `timings` does not live long enough
  --> ash/src/extensions/nv/low_latency2.rs:94:29
   |
93 |         let mut timings = vec![vk::LatencyTimingsFrameReportNV::default()];
   |             ----------- binding `timings` declared here
94 |         info = info.timings(&mut timings);
   |                             ^^^^^^^^^^^^ borrowed value does not live long enough
95 |     }
   |     - `timings` dropped here while still borrowed
96 |     _ = info;
   |         ---- borrow later used here
```

Illegal aliasing:
```
error[E0502]: cannot borrow `timings` as immutable because it is also borrowed as mutable
  --> ash/src/extensions/nv/low_latency2.rs:93:10
   |
92 |     let info = vk::GetLatencyMarkerInfoNV::default().timings(&mut timings);
   |                                                              ------------ mutable borrow occurs here
93 |     dbg!(&timings);
   |          ^^^^^^^^ immutable borrow occurs here
94 |     _ = info;
   |         ---- mutable borrow later used here
```

--- a/ash/src/vk/definitions.rs
+++ b/ash/src/vk/definitions.rs
@@ -52045,9 +52045,9 @@ unsafe impl<'a> TaggedStructure for GetLatencyMarkerInfoNV<'a> {
 }
 impl<'a> GetLatencyMarkerInfoNV<'a> {
     #[inline]
-    pub fn timings(mut self, timings: &'a mut [LatencyTimingsFrameReportNV<'a>]) -> Self {
+    pub fn timings(mut self, timings: &'a mut [LatencyTimingsFrameReportNV<'_>]) -> Self {
         self.timing_count = timings.len() as _;
-        self.p_timings = timings.as_mut_ptr();
+        self.p_timings = timings.as_mut_ptr().cast();
         self
     }
 }

--- a/generator/src/lib.rs
+++ b/generator/src/lib.rs
@@ -1929,7 +1929,7 @@ fn derive_setters(
 
         let deprecated = member.deprecated.as_ref().map(|d| quote!(#d #[allow(deprecated)]));
         let param_ident = field.param_ident();
-        let type_lifetime = has_lifetimes
+        let mut type_lifetime = has_lifetimes
             .contains(&name_to_tokens(&field.basetype))
             .then(|| quote!(<'a>));
         let param_ty_tokens = field.safe_type_tokens(quote!('a), type_lifetime.clone(), None);
@@ -2018,13 +2018,19 @@ fn derive_setters(
         // TODO: Improve in future when https://github.com/rust-lang/rust/issues/53667 is merged id:6
         if field.reference.is_some() && matches!(field.array, Some(vkxml::ArrayType::Dynamic)) {
             if let Some(ref array_size) = field.size {
-                let mut slice_param_ty_tokens = field.safe_type_tokens(quote!('a), type_lifetime.clone(), None);
-
                 let mut ptr = if field.is_const {
                     quote!(.as_ptr())
+                } else if let Some(tl) = &mut type_lifetime {
+                    // Work around invariance with mutable pointers:
+                    // https://github.com/ash-rs/ash/issues/837
+                    // https://doc.rust-lang.org/nomicon/subtyping.html#variance
+                    *tl = quote!(<'_>);
+                    quote!(.as_mut_ptr().cast())
                 } else {
                     quote!(.as_mut_ptr())
                 };
+
+                let mut slice_param_ty_tokens = field.safe_type_tokens(quote!('a), type_lifetime.clone(), None);
 
                 // Interpret void array as byte array
                 if field.basetype == "void" && matches!(field.reference, Some(vkxml::ReferenceType::Pointer)) {

